modal/_container_io_manager.py

```diff
@@ -162,6 +162,86 @@ def validate_output_data(self, data: Any) -> List[Any]:
         return data
 
 
+class ConcurrencyManager:
+    def __init__(self, input_concurrency: int):
+        self._input_concurrency = input_concurrency
+        self._queue = asyncio.Queue()
+        self._owed_releases = 0
+        # lock on access to _input_concurrency and _owed_reductions
+        self._lock = asyncio.Lock()
+
+        for _ in range(input_concurrency):
+            self._queue.put_nowait(None)
+
+    async def get_input_concurrency(self) -> int:
+        async with self._lock:
+            return self._input_concurrency
+
+    async def acquire(self):
+        await self._queue.get()
+
+    async def release(self):
```

Comment:
if you change `async def` to `def` then you don't need a lock since you never yield